spec/src/main/asciidoc/microprofile-lra-spec.adoc

```diff
@@ -727,22 +773,50 @@ annotations when associated with JAX-RS resource methods:
 | 200, 404, 412
 | no expectations
 
+| `@AfterLRA`
+| PUT
+| 200
+| no expectations
+
 |=== 
 
-The presented HTTP methods SHOULD be applied at individual JAX-RS participant methods 
+The presented HTTP methods SHOULD be applied at individual JAX-RS methods
 but they are not enforced by the specification. 
 
+If the method annotated with `@AfterLRA` returns an unexpected HTTP status
+or never reaches the caller then the implementation MUST invoke the same method again.
+
 If the annotated method returns an unexpected HTTP status code the implementation MAY
-invoke the same method again.
+invoke the same method again with the following caveat: if there is no `@Status` method
+and the implementation receives an unexpected response code from either of the
+`@Compensate` or `@Complete` invocations then it MUST reinvoke the method. [Note that
+this caveat applies to the situation where the response is lost since the caller will
+not see the correct code].
 
-Users are allowed to reuse existing JAX-RS endpoints for participant methods definitions. 
+Users are allowed to reuse existing JAX-RS endpoints for participant methods definitions.
 In this case, LRA implementation MUST ensure that invoking these methods outside of the 
 implementation of the LRA specification will not influence any running LRA.
 
+===== Non-JAX-RS afterLRA method
+
+A method annotated with `@AfterLRA` that is not a JAX-RS resource method MUST accept
+two arguments of type URI and LRAStatus, in that order. The first parameter holds
+an LRA context and the second parameter holds the final status of the LRA. If the
+signature does not conform to this requirement then the behaviour of the implementation
+is unspecified.
```

Comment:
This is more a question: we can catch invalid signature during deployment, would it make sense to add another exception for invalid signatures? Something like InvalidAfterLRADefinitionException? I am also ok with currently stated unspecified behavior as it is not participant-specific and thus the LRA protocol does not rely on it.

Reply:
@xstefank I wonder who would catch this InvalidAfterLRADefinitionException?

I think unspecified is fine here, either the vendor might choose to log the error or to fail the deployment or indeed perhaps some other option (I am not sure what exactly)

Reply:
@tomjenkinson indeed, it would be runtime exception failing deployment. I am not sure we want to promise that notification will be delivered but runtime reflective invocation will fail. The business logic may depend on this functionality to start a new LRA for instance.

It would be the same case as for [InvalidLRAParticipantDefinitionException](https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/participant/InvalidLRAParticipantDefinitionException.java)

Reply:
I wonder if we should remove https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/participant/InvalidLRAParticipantDefinitionException.java for the same reason?

Reply:
I would prefer to fail as soon as possible. Why would we allow invalid signatures to be deployed when we are sure they will fail when they are called for the first time?

Reply:
I am not saying a vendor should not be allowed to choose to fail to deploy, but my question is if we have to define the exception in the spec. I.e. couldn't a vendor just throw a RTE and get the same behaviour. Let's take Narayana as an example in something like Thorntail. If we through an RTE("Invalid definition....") won't weld treat it the same way as if we throw this specialised exception?

Reply:
yes, that would be the same behavior. Specifying that only some RTE should be thrown during deployment time would enough. But what is the issue with a custom exception in the specification? Is it a problem that it is a part of the API? If you prefer, I can raise a PR that will remove it.

Reply:
The issue is that we add an unnecessary maintenace element into the spec without a use case - or in other words, I do not think that the additional exception will be treated any differently to a RTE by any caller? Perhaps there are some cases you can think of where a specialised exception would be expected to be caught and dealt with differently from a RTE and if that is so we can use the same argument here?

Reply:
@tomjenkinson I've raised https://github.com/eclipse/microprofile-lra/issues/180

Reply:
Great - thanks @xstefank !

Reply:
I think we should make this consistent with #180 and expect vendors to fail deployment and remove the wording about behaviour being unspecified.

Reply:
So can I resolve this conversation?

Reply:
for me, yes